File: card-image-creator/create-flash.c
#include <stdio.h>
#include <stdlib.h>
#include <errno.h>
#include <string.h>

int yearfuzz = 0; // used for debugging; added to all years (set to -2, e.g.)

const char *event_file = "events.txt";

typedef struct _event_entry {
  char *filename;
  int id;
  int bitpattern;
  int day;
  int month;
  int year;
} event_entry;

typedef struct _media_entry {
  char *filename;
  int local_prefix;
} media_entry;

char *sox_cmd = "sox -v %.2f %s -r 10000 -c 1 -e unsigned-integer -b 16 -t raw %s 2>&1|grep clip >/dev/null";

char *media_prefix = "media";
char local_prefix[255] = "";
char *cache_prefix = "cache";

event_entry events[500] = {
			/* repeating events, like so:
			 {"some-message.wav",
			 0xID, 0x00, 
			 <day>, <month>, 0xFF, 0},
			 ... where 0xID is >= 0xA0 and < 0xF0
			*/

			/* one-time events, like so:
			{"one-time-message.wav",
			0x<ID>, 0x<BIT>,
			<day>, <mon>, <year>, 0 },
			... where 0xID >= 0x10 (and < max ram byte for PIC),
			    and 0x<BIT> is a bit mask in that byte to refer to
			    this specific event (0x01, 0x02, 0x04, 0x08.. 0x80)
			    and 'year' is a two-digit year
			*/


  { NULL, 0, 0, 0, 0 }
};
int num_events = 0;
						
media_entry alarm_song = {"alarm.wav", 0};
float alarm_volume = 0.8;
char *firmware_image = "../main/firmware.bin";
unsigned char firmware_version;

/* The basic media, all of which are included in the media/ directory, 
 * must match their corresponding  IDX_ numbers (in main/constants.inc).
 */
media_entry media[500] = { {"0.wav", 0},
			   {"1.wav", 0},
			   {"2.wav", 0},
			   {"3.wav", 0},
			   {"4.wav", 0},
			   {"5.wav", 0},
			   {"6.wav", 0},
			   {"7.wav", 0},
			   {"8.wav", 0},
			   {"9.wav", 0},
			   {"10.wav", 0},
			   {"11.wav", 0},
			   {"12.wav", 0},
			   {"13.wav", 0},
			   {"14.wav", 0},
			   {"15.wav", 0},
			   {"16.wav", 0},
			   {"17.wav", 0},
			   {"18.wav", 0},
			   {"19.wav", 0},
			   {"20.wav", 0},
			   {"21.wav", 0},
			   {"22.wav", 0},
			   {"23.wav", 0},
			   {"30.wav", 0},
			   {"40.wav", 0},
			   {"50.wav", 0},
			   {"60.wav", 0},
			   {"70.wav", 0},
			   {"80.wav", 0},
			   {"90.wav", 0},
			   {"hundredhours.wav", 0},
			   {"thetimeis.wav", 0},
			   {"setalarmhour.wav", 0},
			   {"setalarm.wav", 0},
			   {"tensofminutes.wav", 0},
			   {"onesofminutes.wav", 0},
			   {"settimehour.wav", 0},
			   {"settime.wav", 0},
			   {"setmonth.wav", 0},
			   {"setday.wav", 0},
			   {"setdow.wav", 0},
			   {"settensofyears.wav", 0},
			   {"setonesofyears.wav", 0},
			   {"timesetto.wav", 0},
			   {"alarmsetto.wav", 0},
			   {"january.wav", 0},
			   {"february.wav", 0},
			   {"march.wav", 0},
			   {"april.wav", 0},
			   {"may.wav", 0},
			   {"june.wav", 0},
			   {"july.wav", 0},
			   {"august.wav", 0},
			   {"september.wav", 0},
			   {"october.wav", 0},
			   {"november.wav", 0},
			   {"december.wav", 0},
			   {"sunday.wav", 0},
			   {"monday.wav", 0},
			   {"tuesday.wav", 0},
			   {"wednesday.wav", 0},
			   {"thursday.wav", 0},
			   {"friday.wav", 0},
			   {"saturday.wav", 0},
			   {"alarmlight.wav", 0},
			   {"alarmstartsearly.wav", 0},
			   {"alarmstartsontime.wav", 0},
			   {"andtodayis.wav", 0},
			   {"blue.wav", 0},
			   {"deadair.wav", 0},
			   {"disabled.wav", 0},
			   {"oclock.wav", 0},
			   {"set.wav", 0},
			   {"setyear.wav", 0},
			   {"white.wav", 0},
		     /* Any media that is listed in the event list above must 
		      * also be listed here */
		     
			   {NULL, 0} };
int num_media = 0;

FILE *output_file;
unsigned long directory_pos = 0;
unsigned long media_pos = 100 * 512;	// byte counter to the start of the media blocks.
unsigned char directory[100*512]; // 100 pages of directory/bootimg space (0-99).

int media_id(const char *filename)
{
  media_entry *p = media;
  int idx = 0;
  while (p->filename) {
    if (!strcmp(p->filename, filename))
      return idx;
    p++;
    idx++;
  }
  printf("ERROR: unable to determine media id for %s\n", filename);
  exit(-2);
}

unsigned char crc7_calc( unsigned char *buff, unsigned long len )
{
  unsigned char crc, i;
  unsigned long j;

  i = 7;// bit position
  j = 0;// array position
  crc = 0;// 7 bit crc value
  while(!((j == len - 1) && (i == 0)) )// crc7
    {
      if((crc = (crc << 1) | ((buff[j] >> i) & 1 )) & 0x80 ) crc ^= 0x89;
      if( i-- == 0 ){ j++; i = 7; 
      }
    }
  return crc << 1;// shift left once
}

// FIXME: file must be < 16k
unsigned char crc7_file(FILE *f)
{
  unsigned char buf[(0x1b00-4)*2];
  size_t s = fread(buf, 1, sizeof(buf), f);
  if (s != sizeof(buf)) {
    fprintf(stderr, "ERROR: unable to read firmware as one chunk (0x%X/0x%X).\n", s,sizeof(buf));
    fprintf(stderr, "       Rewrite crc7_file to be more sane.\n");
    exit(-6);
  }
  return crc7_calc(buf, sizeof(buf));
}

unsigned char to_bcd(unsigned char v)
{
    return (( v / 10 ) * 16) + (v % 10);
}


// add the given media clip to the flash image, and return the length. We have to do some 
// processing of the media clip in order to make it playable; the PIC firmware will dump the 
// raw bytes down the D/A channel, so we have to prepend command information to the bytes 
// here.
unsigned long add_clip(media_entry *m, float initial_volume)
{
	size_t s;
	unsigned char buf[512];
	unsigned long length = 0;
	char fullpath[1024], cachepath[1024], cmdbuf[2048];
	FILE *f;
	unsigned long i;
	float volume = initial_volume;
	if (volume == 0)
	  volume = 5.0;

	sprintf(fullpath, "%s/%s", m->local_prefix ? local_prefix : media_prefix, m->filename);
	sprintf(cachepath, "%s/%s", cache_prefix, m->filename);

	f = fopen(fullpath, "r");
	if (!f) {
		fprintf(stderr, "Failed to open %s: %d\n", fullpath, errno);
		exit(-1);
	}
	fclose(f);

	// adjust the volume of the clip - as loud as possible without distortion!
	f = fopen(cachepath, "r");
	if (f) {
	  close(f);
	  goto got_one_already;
	}

	while (1) {
	  int res;

	  sprintf(cmdbuf, sox_cmd, volume, fullpath, cachepath);
	  res = system(cmdbuf);
	  if (res == 256) 
	    break;
	  volume -= 0.1;
	  if (volume <= 0.1) {
	    fprintf(stderr, "failed.\n");
	    exit(-6);
	  }
	  printf(" - processed at volume %.2f\n", volume);
	}
 got_one_already:
	f = fopen(cachepath, "r");
	if (!f) {
	  fprintf(stderr, "Failed to open tmp path %s: %d\n", cachepath, errno);
	  exit(-1);
	}
	printf(" - starting at media pos %lu\n", media_pos);
	fseek(output_file, media_pos, SEEK_SET);
	while (1) {
		s = fread(buf, 1, 512, f);
		if (!s)
			break;
		for (i=0; i<s; i+=2) {
			unsigned long val = (buf[i+1] << 8) + buf[i];
			val >>= 4; // make it a 12-bit value instead of 16
			val |= 0x3000; // prepend D/A command header (%0011xxxx)
			// put the data into a media block, big-endian
			fputc((val >> 8) & 0xFF, output_file);
			fputc((val     ) & 0xFF, output_file);
			media_pos += 2;
			length += 2;
		}
	}
	fclose(f);
	// pad the media block
	while (media_pos % 512) {
		// push on "dead air" (header + median value)
	  fputc(0x38, output_file);
	  fputc(0x00, output_file);
	  media_pos += 2;
	  length += 2;
	}

	return length;
}

void construct_events()
{
	unsigned long s, start_block;
	event_entry *e;
	
	// first event is the alarm. It's special.
	directory_pos = 0; // start of block 0, which is where events live.

	printf("Constructing alarm event from %s\n", alarm_song.filename);
	// add the directory entry for the alarm
	start_block = media_pos / 512;
	directory[directory_pos++] = (start_block >> 24) & 0xFF;
	directory[directory_pos++] = (start_block >> 16) & 0xFF;
	directory[directory_pos++] = (start_block >>  8) & 0xFF;
	directory[directory_pos++] = (start_block      ) & 0xFF;
	s = add_clip(&alarm_song, alarm_volume);
	start_block = media_pos / 512;
	directory[directory_pos++] = (start_block >> 24) & 0xFF;
	directory[directory_pos++] = (start_block >> 16) & 0xFF;
	directory[directory_pos++] = (start_block >>  8) & 0xFF;
	directory[directory_pos++] = (start_block      ) & 0xFF;
	
	// add directory entries for all of the "real" events. These take a 
	// filename in the struct, but they use that to determine the media 
	// id in the media queue.
	e = events;
	while (e && e->filename) {
	  if (e->year == 0xFF) {
	    if (e->bitpattern) {
	      printf("Processing event id %.2X/%.2X for %d/%d every year from %s\n",
		     e->id, e->bitpattern,
		     e->month, e->day, 
		     e->filename);
	    } else {
	      printf("Processing transient event %.2X for %d/%d every year from %s\n",
		     e->id,
		     e->month, e->day, 
		     e->filename );
	    }
	  } else {
	    if (e->bitpattern) {
	      printf("Processing event id %.2X/%.2X for %d/%d/%d from %s\n",
		     e->id, e->bitpattern,
		     e->month, e->day, 2000+e->year+yearfuzz,
		     e->filename);
	    } else {
	      printf("Processing transient event id %.2X for %d/%d/%d from %s\n",
		     e->id,
		     e->month, e->day, 2000+e->year,
		     e->filename);
	    }
	  }
		// 2 bytes of ID
		directory[directory_pos++] = e->id;
		directory[directory_pos++] = e->bitpattern;
		
		// year, month, day in BCD.
		if (e->year == 0xFF) 
		  directory[directory_pos++] = e->year;
		else
		  directory[directory_pos++] = to_bcd(e->year+yearfuzz);
		directory[directory_pos++] = to_bcd(e->month);
		directory[directory_pos++] = to_bcd(e->day);

		// media id.
		directory[directory_pos++] = media_id(e->filename);
		
		e++;
	}

	// 6 bytes per event.  8 bytes eaten up by alarm header.
	int num_events = ( ((3 * 512) - 8) / 6 );

	if (directory_pos >= 3*512) {
	  // Room for 3 pages of events (0-2), minus 8 bytes of alarm 
	  // header information, divided by the size of one event in 
	  // the directory (13 bytes).
	  fprintf(stderr, "ERROR: event directory is too long (room for %d events + 1 alarm).\n", num_events);
	  exit(-2);
	}

	printf("event directory has room for %d events + 1 alarm.\n", num_events);
}

void construct_media()
{
        media_entry *p;
	int idx = 0;
	unsigned long s;
	unsigned long start_block;
	
	// Page 3 of directory: media files for time setting.
	directory_pos = 3*512;

	p = &media[idx++];
	while (p->filename) {
	  printf("Processing media file %s\n", p->filename);
		start_block = media_pos / 512;
		directory[directory_pos++] = (start_block >> 24) & 0xFF;
		directory[directory_pos++] = (start_block >> 16) & 0xFF;
		directory[directory_pos++] = (start_block >>  8) & 0xFF;
		directory[directory_pos++] = (start_block      ) & 0xFF;
		s = add_clip(p, 0);
		printf(" -- clip is %lu bytes long\n", s);
		start_block = media_pos / 512;
		directory[directory_pos++] = (start_block >> 24) & 0xFF;
		directory[directory_pos++] = (start_block >> 16) & 0xFF;
		directory[directory_pos++] = (start_block >>  8) & 0xFF;
		directory[directory_pos++] = (start_block      ) & 0xFF;
	
		p = &media[idx++];
	}

	if (1) {
	  int num_events = ( ((10-3) * 512) / 8 );
	  printf("Added %d pieces of media (room for %d; can only address 256 in code though).\n", idx-1, num_events);
	}

	if (directory_pos >= 10*512) {
	  fprintf(stderr, "ERROR: media directory is too long.\n");
	  exit(-3);
	}
}

void construct_firmware()
{
	FILE *f;
	unsigned long length = 0;
	unsigned char buf[512];
	unsigned long i;
	unsigned char crc7;

	// Page #10 (the 11th) of directory: firmware upgrade.
	directory_pos = 10*512;
	
	printf("Processing firmware version 0x%.2X from %s\n", firmware_version, firmware_image);
	
	f = fopen(firmware_image, "r");
	if (!f) {
	  fprintf(stderr, "Unable to open firmware image: %d\n", errno);
	  exit(-1);
	}

	directory_pos += 3; // skip 3 bytes for version#, size
	crc7 = crc7_file(f);
	printf("checksum: 0x%.2X\n", crc7);
	directory[directory_pos++] = crc7;
	fseek(f, 0, SEEK_SET);

	while (1) {
		size_t s = fread(buf, 1, 512, f);
		if (!s)
			break;
		for (i=0; i<s; i+=2) {
			unsigned short val = (buf[i] << 8) + buf[i+1];
			// put the data into a directory, little-endian
			directory[directory_pos++] = (val     ) & 0xFF;
			directory[directory_pos++] = (val >> 8) & 0xFF;
			length += 2;
		}
	}
	fclose(f);

	directory_pos = 10*512; // rewind...
	directory[directory_pos++] = firmware_version;
	length = 0x1e00 * 2; // FIXME: hack, trying to figure out what's up with firmware problems
	length /= 2; // need length in *words*, not *bytes*
	directory[directory_pos++] = (length >> 8) & 0xFF;
	directory[directory_pos++] = (length     ) & 0xFF;
}

void write_directory()
{
	fseek(output_file, 0, SEEK_SET);
	fwrite(directory, sizeof(directory), 1, output_file);
}

void count_events()
{
  num_events = 0;
  while (events[num_events].filename) {
    num_events++;
  }
}

void count_media()
{
  num_media = 0;
  while (media[num_media].filename) {
    num_media++;
  }
}

// read list of events from event_file, adding them to media[] and events[]
void read_events()
{
  FILE *f = fopen(event_file, "r");
  if (!f)
    return;

  char line[255];
  char *p;

  char filename[255];
  float volume;
  int byte, bitmask, day, month, year;

  while (p = fgets(line, sizeof(line), f)) {

    if (sscanf(p, "prefix %s", filename) == 1) {
      strcpy(local_prefix, filename);
    } else if (sscanf(p, "alarm %s %f", filename, &volume) == 2) {
      alarm_song.filename = malloc(strlen(filename)+1);
      strcpy(alarm_song.filename, filename);
      alarm_song.local_prefix = 1;
      alarm_volume = volume;

    } else if (sscanf(p, "%s %d %d %d %d %d\n", filename, &byte, &bitmask, &day, &month, &year) == 6) {
      event_entry *e = &events[num_events];
      e->filename = malloc(strlen(filename)+1);
      strcpy(e->filename, filename);
      e->id = byte;
      e->bitpattern = bitmask;
      e->day = day;
      e->month = month;
      e->year = year;
      media[num_media].filename = e->filename;
      media[num_media].local_prefix = 1;

      e = &events[num_events+1];
      memset(e, 0, sizeof(event_entry));

      num_events++;
      num_media++;
      media[num_media].filename = NULL;
    } else if (strncmp(p, "//", 2) == 0) {
      // Ignore the comment...
    } else {
      fprintf(stderr, "ERROR: failed to parse line: '%s'\n", p);
      exit(-11);
    }
  }

  fclose(f);
}

int main(int argc, char *argv[])
{
	memset(directory, 0, sizeof(directory));

	count_events();
	count_media();

	read_events();

	firmware_version = system("./version.pl")>>8;
	output_file = fopen("flash.img", "w");
	if (!output_file) {
		fprintf(stderr, "Unable to open output image: %d", errno);
		exit(-1);
	}
	
	construct_events();
	construct_media();
	construct_firmware();
	
	write_directory();
	
	fclose(output_file);

	printf("... done. Generated firmware is named \"flash.img\".\n");
	
	return 0;
}

File: card-image-creator/create-flash.pl
#!/usr/bin/perl

use strict;
use warnings;
use Fcntl qw(:DEFAULT :seek);
use bigint lib => 'Calc';

# First 4 bytes: start/end of the alarm event.
# Following that are blocks:
#  -> 1 byte, 1 bitmask for the event;
#  -> 4 bytes, start block of sound;
#  -> 4 bytes, end block of sound.
# The last event is all 0s.
#
# Each block is 512 bytes long.
#
# Block 0 contains the directory for alarm events.
# Block 1 contains the directory of parts-of-time-speech.
# Block 2 contains the flash version, if any, that's on this card.
#   If there is a flash version, blocks 3+ might also contain flash data.
# All other blocks are data.

$|=1;

my $ofh;
open($ofh, ">output.img") || die;
my $mapfile;
open($mapfile, ">mapfile.dat") || die;

# Configurables
my $start_id = 0x10;
my $alarm_song = 'chicago.wav2.raw';
#my $alarm_song = 'jake-hello.wav.raw';
#my @songs = ( 'boogie.wav2.raw', 'whels.wav2.raw', 'yoyo.wav2.raw' );
my @songs = ( '1.wav.raw', '2.wav.raw', '3.wav.raw' );
my %songdata = ( 'boogie.wav2.raw' => { day => 19, 
					month => 4,
					year => 10 },
		 'whels.wav2.raw' => { day => 20,
				       month => 4,
				       year => 10 },
		 'yoyo.wav2.raw' => {day => 21,
				     month => 4,
				     year => 10 },
		 '1.wav.raw' => { day => 7, 
				  month => 5,
				  year => 9 },
		 '2.wav.raw' => { day => 8, 
				  month => 5,
				  year => 9 },
		 '3.wav.raw' => { day => 9,
				  month => 5,
				  year => 9 },
);

my @media = ( '0.wav.raw', 
	      '1.wav.raw', 
	      '2.wav.raw', 
	      '3.wav.raw', 
	      '4.wav.raw', 
	      '5.wav.raw', 
	      '6.wav.raw', 
	      '7.wav.raw', 
	      '8.wav.raw', 
	      '9.wav.raw', 
	      '10.wav.raw', 
	      '11.wav.raw', 
	      '12.wav.raw', 
	      '13.wav.raw', 
	      '14.wav.raw', 
	      '15.wav.raw', 
	      '16.wav.raw', 
	      '17.wav.raw', 
	      '18.wav.raw', 
	      '19.wav.raw', 
	      '20.wav.raw', 
	      '21.wav.raw', 
	      '22.wav.raw', 
	      '23.wav.raw', 
	      '30.wav.raw',
	      '40.wav.raw',
	      '50.wav.raw',
	      '60.wav.raw',
	      '70.wav.raw',
	      '80.wav.raw',
	      '90.wav.raw',
	      'hundredhours.wav.raw',
	      'timeis.wav.raw',
	      # Second set of media clips: time/alarm setting info
	      'setalarmhour.wav.raw',
	      'setalarmminute.wav.raw',
	      'settimehour.wav.raw',
	      'settimeminute.wav.raw',
	      'setmonth.wav.raw',
	      'setday.wav.raw',
	      'setdow.wav.raw',
	      'tensyears.wav.raw',
	      'onesyears.wav.raw',
	      'timesetto.wav.raw',
	      'alarmsetto.wav.raw',
	      'january.wav.raw',
	      'february.wav.raw',
	      'march.wav.raw',
	      'april.wav.raw',
	      'may.wav.raw',
	      'june.wav.raw',
	      'july.wav.raw',
	      'august.wav.raw',
	      'september.wav.raw',
	      'october.wav.raw',
	      'november.wav.raw',
	      'december.wav.raw',
	      'sunday.wav.raw',
	      'monday.wav.raw',
	      'tuesday.wav.raw',
	      'wednesday.wav.raw',
	      'thursday.wav.raw',
	      'friday.wav.raw',
	      'saturday.wav.raw'
 );
my @media_directory;
my $media_dir_pos = 0;

# end configurables

my @directory;

while ($#directory != 511) {
    push (@directory, 0);
}
while ($#media_directory != 511) {
    push (@media_directory, 0);
}

my $start_block = 33; # where media starts being encoded to.
print $mapfile "Block $start_block";

my (@encoding) = process_song($alarm_song);
print "Encoded length: " , $#encoding + 1, "\n";
my $directory_pos = 0;
$directory[$directory_pos++] = ($start_block >> 24) & 0xFF;
$directory[$directory_pos++] = ($start_block >> 16) & 0xFF;
$directory[$directory_pos++] = ($start_block >> 8) & 0xFF;
$directory[$directory_pos++] = ($start_block ) & 0xFF;
$start_block += ($#encoding + 1) / 512;
$directory[$directory_pos++] = ($start_block >> 24) & 0xFF;
$directory[$directory_pos++] = ($start_block >> 16) & 0xFF;
$directory[$directory_pos++] = ($start_block >> 8) & 0xFF;
$directory[$directory_pos++] = $start_block & 0xFF;
print "New start block: $start_block\n";
print $mapfile "Block $start_block";

# Dump the directory and alarm song to the output image.
foreach my $i (@directory) {
    print $ofh sprintf("%c", $i);
}
# Dump dummy space for blocks 1-32, which we'll skip so there's room for a 
# new firmware image...
foreach my $i (1..32) {
    foreach my $j (0..511) {
	print $ofh (sprintf "%c", 0);
    }
}

foreach my $i (@encoding) {
    print $ofh sprintf("%c", $i);
}
@encoding = undef;


# Now add events for some more songs. Start @ id #1; #0 is 
# reserved for "there are no more events"
my (@id) = ($start_id, 0x01); # EEPROM id, bit flag in that block

foreach my $song (@songs) {
    @encoding = process_song($song);
    print "Encoded length: " , $#encoding + 1, "\n";
    foreach my $i (@encoding) {
	print $ofh sprintf("%c", $i);
    }

    add_directory(\@directory, \$directory_pos, \@id, \$start_block, 
		  \@encoding,
		  to_bcd($songdata{$song}->{day}),
		  to_bcd($songdata{$song}->{month}),
		  to_bcd($songdata{$song}->{year}) );

    @encoding = undef;
    print "New start block: $start_block\n";
    print $mapfile "Block $start_block";

    $id[1] <<= 1;
    if ($id[1] >= 0x100) {
	$id[0]++;
	$id[1] = 1;
	print "Starting on ID byte $id[0]\n";
    }
}

# Now take each of the media files and dump 'em. Build the media directory.
foreach my $media (@media) {
    @encoding = process_song($media);
    print "Encodedmedia length: " , $#encoding + 1, "\n";
    foreach my $i (@encoding) {
	print $ofh sprintf("%c", $i);
    }

    add_media(\@media_directory, \$media_dir_pos, \$start_block, \@encoding);

    @encoding = undef;
    print $mapfile "Block $start_block";
}

print $mapfile "\n";
# Okay, all done adding - dump the final directory and media directory.
seek($ofh, 0, SEEK_SET);
foreach my $i (@directory) {
    print $ofh sprintf("%c", $i);
}
foreach my $i (@media_directory) {
    print $ofh sprintf("%c", $i);
}

embed_firmware($ofh);

close $ofh;
exit;

sub embed_firmware
{
    my ($ofh) = @_;
  # Now embed any firmware we have. Write it big-endian. The original file is,
  # of course, little-endian. So swap.
    seek($ofh, 2 * 512, SEEK_SET); # Firmware goes on page 2 (the third page).
    my $FIRMWARE_VERSION = 4; # FIXME
    print $ofh sprintf("%c", $FIRMWARE_VERSION);
    my $bytecount = 7680; # FIXME: make more dynamic somehow?
    print $ofh sprintf("%c%c", int($bytecount/256), $bytecount%256);
    print "writing firmware... ";
    my $fw;
    open($fw, 'firmware.dat') || die;
    while (1) {
	my $b;
	my $size = sysread($fw, $b, 1024*1024);
	(print "\n" && last)
	    if ($size == 0);
	print ".";
	
	# the raw data is low byte first, of course. Grr.
	my (@a) = unpack('v*', $b);
	foreach my $a (@a) {
	    print $ofh sprintf("%c%c", int($a/256), $a%256);
	}
    }
    close $fw;
}

sub to_bcd {
    my ($val) = @_;

    # Convert from decimal to BCD. Assume it's < 100 (Decimal).
    return (int( $val / 10 ) * 16) + ($val % 10);
}

sub add_media {
    my ($dir, $dpos, $startb, $enc) = @_;

    my @e = @$enc;
    $dir->[$$dpos++] = ($$startb >> 24) & 0xFF;
    $dir->[$$dpos++] = ($$startb >> 16) & 0xFF;
    $dir->[$$dpos++] = ($$startb >> 8) & 0xFF;
    $dir->[$$dpos++] = ($$startb) & 0xFF;
    $$startb += ($#e + 1) / 512;
    $dir->[$$dpos++] = ($$startb >> 24) & 0xFF;
    $dir->[$$dpos++] = ($$startb >> 16) & 0xFF;
    $dir->[$$dpos++] = ($$startb >> 8) & 0xFF;
    $dir->[$$dpos++] = $$startb & 0xFF;
}

sub add_directory {
    my ($dir, $dpos, $id, $startb, $enc, $start_day, $start_month, $start_year) = @_;

    my @e = @$enc;

    $dir->[$$dpos++] = $id->[0];
    $dir->[$$dpos++] = $id->[1];
    $dir->[$$dpos++] = $start_year;
    $dir->[$$dpos++] = $start_month;
    $dir->[$$dpos++] = $start_day;
    $dir->[$$dpos++] = ($$startb >> 24) & 0xFF;
    $dir->[$$dpos++] = ($$startb >> 16) & 0xFF;
    $dir->[$$dpos++] = ($$startb >> 8) & 0xFF;
    $dir->[$$dpos++] = ($$startb) & 0xFF;
    $$startb += ($#e + 1) / 512;
    $dir->[$$dpos++] = ($$startb >> 24) & 0xFF;
    $dir->[$$dpos++] = ($$startb >> 16) & 0xFF;
    $dir->[$$dpos++] = ($$startb >> 8) & 0xFF;
    $dir->[$$dpos++] = $$startb & 0xFF;
}

sub process_song {
    my ($fn) = @_;

    print "Processing $fn\n";
    print $mapfile ": $fn\n";

    my $fh;
    open($fh, $fn) || die;

    my @ret;
    
    while (1) {
	my $b;
	my $size = sysread($fh, $b, 1024*1024);
	(print "\n" && last)
	    if ($size == 0);

	print ".";

	# the raw data is low byte first, of course. Grr.
	my (@a) = unpack('v*', $b);
#	print "have ", $#a + 1, " bytes\n";

	foreach my $a (@a) {
	    # Drop 4 bits off the top.
	    $a >>= 4;
	    
	    # Add the D/A header (%0011xxxx).
	    $a |= 0x3000;
	    
	    # Save the new value in big-endian order.
	    push (@ret, unpack('C*', pack('n', $a)));
	}
    }

    print "padding $fn\n";

    while (($#ret + 1) % 512) {
	# SPI device ID header (%0011) and a median PWM value (%1000...).
	push(@ret, 0x38, 0x00);
    }

    return @ret;
}

